src/mlpack/methods/reinforcement_learning/q_networks/simple_dqn.hpp

```diff
@@ -48,16 +48,14 @@ class SimpleDQN
   SimpleDQN(const int inputDim,
             const int h1,
             const int h2,
-            const int outputDim) : network()
+            const int outputDim) : network(MeanSquaredError<>(),
+                GaussianInitialization(0, 0.001))
   {
-    FFN<MeanSquaredError<>, GaussianInitialization> model(MeanSquaredError<>(),
-        GaussianInitialization(0, 0.001));
-    model.Add<Linear<>>(inputDim, h1);
-    model.Add<ReLULayer<>>();
-    model.Add<Linear<>>(h1, h2);
-    model.Add<ReLULayer<>>();
-    model.Add<Linear<>>(h2, outputDim);
-    network = model;
+    network.Add<Linear<>>(inputDim, h1);
+    network.Add<ReLULayer<>>();
+    network.Add<Linear<>>(h1, h2);
+    network.Add<ReLULayer<>>();
+    network.Add<Linear<>>(h2, outputDim);
```

Comment:
@zoq Shouldn't this work? And yet somehow its getting attached to [this](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/add.hpp) `Add`instead of [this](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn.hpp#L290). 

Obviously, I'm doing something wrong

Reply:
Hey @nishantkr18, There are a bit more changes in this file. You can refer to #2387. If you want to make the changes here that's great.

Reply:
@kartikdutt18 I think the same error is coming up in #2387. `error: invalid use of ‘class mlpack::ann::Add<mlpack::ann::Linear<arma::Mat<double>, arma::Mat<double>, mlpack::ann::NoRegularizer> >’
   network.Add<Linear<>>(h1, h2);`

Reply:
Right, For some they only occur in linux builds. I read a bit more and made changes accordingly to no avail.

Reply:
Interesting, it should work yes, can you test: `Linear<>* linear = new Linear<>(...); model.Add(linear);`?

Reply:
> Interesting, it should work yes, can you test: Linear<>* linear = new Linear<>(...); model.Add(linear);?

This should work as that's what I have used in models repo. Let me give it a try. Thanks a lot.

Reply:
Yeah, it works! :smiley:

Reply:
Still, will see if I can figure out what the problem is.

Reply:
Btw, I'm facing a problem here:
```
    Concat<> concat = new Concat<>();
    concat.Add<Sequential<>>(valueNetwork);
    concat.Add<Sequential<>>(advantageNetwork);

    featureNetwork.Add(new Linear<>(inputDim, h1));
    featureNetwork.Add(new ReLULayer<>());
    featureNetwork.Add<Concat<>>(concat);
```
Its the same problem, somehow it calls `mlpack::ann::Add` at `featureNetwork.Add<Concat<>>(concat);`

I tried using `Concat<>* concat = new Concat<>();`, it causes problems in forward and backward Pass for the network.

Reply:
Hey @nishantkr18, I think they are at least two ways to concat two or more networks. The first one you tried, I am not sure about the problem you faced using `Concat<>*` , The other if is using boost visitors. I think this is the best implementation of concatenation of [networks](https://github.com/mlpack/mlpack/pull/1950#issuecomment-511765865) using visitors. Let me know what you think about either implementation. 
Also if we don't use mlpack::ann namespace is the issue resolved because then it won't find Add layer?

Reply:
```c++
   Sequential<>* valueNetwork = new Sequential<>();
   valueNetwork->Add(new NoisyLinear<>(32, 16));
   valueNetwork->Add(new ReLULayer<>());
   valueNetwork->Add(new Linear<>(16, 1));

   Sequential<>* advantageNetwork = new Sequential<>();
   advantageNetwork->Add(new Linear<>(32, 16));
   advantageNetwork->Add(new ReLULayer<>());
   advantageNetwork->Add(new Linear<>(16, 2));

   Concat<>* concat = new Concat<>();
   concat->Add<Sequential<>>(valueNetwork);
   concat->Add<Sequential<>>(advantageNetwork);

   FFN<EmptyLoss<>, GaussianInitialization>* featureNetwork =
       new FFN<EmptyLoss<>, GaussianInitialization>(EmptyLoss<>(), GaussianInitialization(0, 0.001));

   featureNetwork->Add<Linear<>>(4, 32);
   featureNetwork->Add<ReLULayer<>>();
   featureNetwork->Add(concat);

   arma::mat input = arma::ones(4, 1);

   featureNetwork->ResetParameters();

   arma::mat output, grad;
   featureNetwork->Forward(input, output);
```

The above snippet compiles with no errors, but is not able to run. I'm not sure what the error is..  

> Also if we don't use mlpack::ann namespace is the issue resolved because then it won't find Add layer?

Yeah, that could be done, but we'd need to add mlpack::ann to classes like `FFN`, `EmptyLoss`, etc.. That might cause unnecessary long lines of code, would that be ok?

Reply:
> Yeah, that could be done, but we'd need to add mlpack::ann to classes like FFN, EmptyLoss, etc.. That might cause unnecessary long lines of code, would that be ok?

You could use namespace mlpack and not mlpack::ann. I think @zoq, would provide better insight here.

Reply:
Agreed, also we could create an alias (`using` or `typedef`).

Reply:
> The above snippet compiles with no errors, but is not able to run. I'm not sure what the error is..

Do you get any errors? Also when you build mlpack make sure to build with `-DDEBUG=ON` to enable the bounds check. My first guess is that the dimension of the last layer has to be the same.

Reply:
> My first guess is that the dimension of the last layer has to be the same.

I'm not sure how that can be done, given that the value network needs to give out a scalar, whereas the advantage function gives out the adv value for each action.. Maybe I didn't get u. 

Anyways, I guess I found the sol, although its just a workaround I guess. I seems that everything works well, after removing [this Add](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/concat.hpp#L141-L147) from Concat.

Reply:
> Anyways, I guess I found the sol, although its just a workaround I guess. I seems that everything works well, after removing this Add from Concat.

I understand that it works but are we sure that it doesn't remove any functionality from mlpack. Like mlpack had it in previous releases if this is merged and a user was using this function he won't be able to use it in any further releases. Maybe I am missing something. Regards.

Reply:
Yeah, I completely get that. Thats why I said its a workaround.. I couldn't find a way to call [this](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/concat.hpp#L157-L162), with [this](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/concat.hpp#L141-L147) present. 
```
    Concat<>* concat = new Concat<>(true);
    concat->Add(valueNetwork);
    concat->Add(advantageNetwork);
```
Maybe @zoq could help?

Reply:
That's fine to remove, because even if that was part of mlpack before, I think you couldn't use it with the interface.

Also what I meant was, that `valueNetwork->Add(new Linear<>(h2, 1));` so as you pointed out it's a matrix of size 1x1, but the `advantageNetwork->Add(new Linear<>(h2, outputDim));` has another size, so maybe the `join_cols` in the forward pass isn't right. But it sounded like you figured it out.

Reply:
>That's fine to remove, because even if that was part of mlpack before, I think you couldn't use it with the interface.

Hey @zoq, I might be missing something but I think that works on OSX and Windows and fails only on Linux Machines. Also if anyone wasn't using the namespace then also it would work irrespective of the machine. Kindly let me know what you think. Thanks.

Reply:
I see, let's remove it for now, so that we can continue with this.

Reply:
Sure, that makes sense. Thanks for the clarification.